src/analyses/cfg_dominators.h

```diff
@@ -44,180 +187,276 @@ class cfg_dominators_templatet
+      node_indext s = w;
+      while (semi[label[w]] < semi[label[child[s]]]) {
+        if (size[s] + size[child[child[s]]] >= 2 * size[child[s]]) {
+          ancestor[child[s]] = s;
+          child[s] = child[child[s]];
+        } else {
+          size[child[s]] = size[s];
+          s = ancestor[s] = child[s];
         }
       }
+      label[s] = label[w];
+      size[v] = size[v] + size[w];
+      if (size[v] < 2 * size[w]) {
+        std::swap(s, child[v]);
+      }
+      while (s != 0) {
+        ancestor[s] = v;
+        s = child[s];
+      }
+    }
 
-      while(n_it!=node.dominators.end())
-      {
-        if(*n_it==current)
-          ++n_it;
-        else
-        {
-          changed=true;
-          node.dominators.erase(n_it++);
+    void assign_dominators(node_indext root) {
+      std::stack<node_indext> work;
+      work.push(root);
+      while (!work.empty()) {
+        node_indext v = work.top();
+        work.pop();
+        if (cfg_dominators.cfg[v - 1].dominators.empty()) {
+          cfg_dominators.cfg[v - 1].dominators =
+              target_sett(cfg_dominators.dominators_data, v - 1);
+          for_each_successor(v, [&](node_indext w) { work.push(w); });
         }
       }
     }
 
-    if(changed) // fixed point for node reached?
-    {
-      for(const auto &edge : (post_dom ? node.in : node.out))
-      {
-        worklist.push_back(cfg[edge.first].PC);
+    bool is_unreachable(node_indext index) {
+      return semi[index] == 0;
+    }
+
+    template<typename Action>
+    void for_each_successor(node_indext node_index, Action action) {
+      // the -1 / +1 adjusts indices from 1 based to 0 based and back
+      auto ix = node_index - 1;
+      for (auto const &next :
+          post_dom ? cfg_dominators.cfg.in(ix) : cfg_dominators.cfg.out(ix)) {
+        action(next.first + 1);
+      }
+    }
+
+    template<typename Action>
+    void for_each_predecessor(node_indext node_index, Action action) {
+      auto ix = node_index - 1;
+      for (auto const &prev :
+          post_dom ? cfg_dominators.cfg.out(ix) : cfg_dominators.cfg.in(ix)) {
+        action(prev.first + 1);
+      }
+    }
+  };
+};
+
+template<class P, class T, bool post_dom>
+void cfg_dominators_templatet<P, T, post_dom>::fixedpointt::fixedpoint(
+    P &program) {
+  // Dominator Tree according to Lengauer and Tarjan
+  // "A fast algorithm for finding dominators in a flow graph"
+  // This is ununderstandable without reading the paper!
+  // assumption: Vertex indices >= 0 and < cfg.size()
+  if (cfg_dominators.cfg.nodes_empty(program)) {
+    return;
+  }
+  cfg_dominators.entry_node = get_entry_node(program);
+  // algorithm also has "pred", for which we can use
+  // cfg.in instead because we already compute
+  // predecessors
+
+  node_indext root =
+      cfg_dominators.cfg.entry_map[cfg_dominators.entry_node] + 1;
+  dfsCounter = 0;
+  dfs(root);
+  for (node_indext i = dfsCounter; i >= 2; --i) {
```

Comment:
Why is it `2`? I'm mainly asking as the original paper might assume indices starting from 1 (and I don't understand the code so I can't tell what it is supposed to do).